compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -478,6 +478,11 @@ object Types {
       case _ => f(this)
     }
 
+    final def mapReduceAndOr[T](f: Type => T)(g: (T, T) => T)(using Context): T = stripTypeVar match {
```

Comment:
That's logically wrong. The base classes of A & B are the union of the base classes of A and the base classes of B not the intersection. 

Is the problem is also there if the union is with a type other than `Null`?

Reply:
Then I believe the fix should also be null-specific. Maybe reassign to @noti0na1 ?

Reply:
I will take a look today